test/blackbox/common/DDSBlackboxTestsDataWriter.cpp

```diff
@@ -533,6 +534,314 @@ TEST(DDSDataWriter, datawriter_qos_use_topic_qos)
+ */
+TEST(DDSDataWriter, datawriter_get_matched_subscriptions_ok_empty_list)
+{
+    using InstanceHandle_t = eprosima::fastrtps::rtps::InstanceHandle_t;
+
+    PubSubWriter<HelloWorldPubSubType> writer(TEST_TOPIC_NAME);
+    PubSubReader<HelloWorldPubSubType> reader_1(TEST_TOPIC_NAME);
+    PubSubReader<HelloWorldPubSubType> reader_2(TEST_TOPIC_NAME);
+
+    std::vector<InstanceHandle_t> sub_handles;
+
+    writer.reliability(BEST_EFFORT_RELIABILITY_QOS)
+            .init();
+
+    reader_1.reliability(RELIABLE_RELIABILITY_QOS)
+            .init();
+
+    reader_2.ownership_exclusive()
+            .init();
+
+    ASSERT_TRUE(writer.isInitialized());
+    ASSERT_TRUE(reader_1.isInitialized());
+    ASSERT_TRUE(reader_2.isInitialized());
+
+    // Writer should not be matched with any reader
+    writer.wait_discovery(2, std::chrono::seconds(2));
+    ASSERT_FALSE(writer.is_matched());
+
+    auto& native_writer = writer.get_native_writer();
+    ReturnCode_t ret = native_writer.get_matched_subscriptions(sub_handles);
+
+    ASSERT_EQ(ret, ReturnCode_t::RETCODE_OK);
+    ASSERT_EQ(sub_handles.size(), 0u);
+}
+
+/**
+ * Refers to DDS-DW-API-GMS-02 from the test plan.
+ *
+ * get_matched_subscriptions() must provide the correct list of matched subscription handles.
+ * Parameterize the test for different transports.
+ */
+TEST_P(DDSDataWriter, datawriter_get_matched_subscriptions_correctly_behaves)
+{
+    using InstanceHandle_t = eprosima::fastrtps::rtps::InstanceHandle_t;
+
+    const size_t num_readers = 5;
+
+    PubSubWriter<HelloWorldPubSubType> writer(TEST_TOPIC_NAME);
+    std::vector<std::unique_ptr<PubSubReader<HelloWorldPubSubType>>> readers;
+    std::vector<InstanceHandle_t> expected_sub_handles;
+    std::vector<InstanceHandle_t> sub_handles;
+
+    readers.reserve(num_readers);
+    sub_handles.reserve(num_readers);
+
+    writer.init();
+
+    ASSERT_TRUE(writer.isInitialized());
+
+    for (size_t i = 0; i < num_readers; ++i)
+    {
+        readers.emplace_back(new PubSubReader<HelloWorldPubSubType>(TEST_TOPIC_NAME));
+        readers.back()->init();
+        ASSERT_TRUE(readers.back()->isInitialized());
+        expected_sub_handles.emplace_back(readers.back()->get_native_reader().get_instance_handle());
+    }
+
+    // Wait for discovery
+    writer.wait_discovery(num_readers, std::chrono::seconds::zero());
+    ASSERT_EQ(writer.get_matched(), num_readers);
+
+    auto& native_writer = writer.get_native_writer();
+    ReturnCode_t ret = native_writer.get_matched_subscriptions(sub_handles);
+
+    // Check that the list of matched publication handles is correct
+    ASSERT_EQ(ret, ReturnCode_t::RETCODE_OK);
+    ASSERT_EQ(sub_handles.size(), num_readers);
+    ASSERT_TRUE(std::is_permutation(sub_handles.begin(), sub_handles.end(), expected_sub_handles.begin()));
+
+    // Remove two readers and check that the list of matched publication handles is updated
+    readers.pop_back();
```

Comment:
I think this will not destroy the readers. You have to use `delete`.

Reply:
Same as above